src/non_interactive.rs

```diff
@@ -152,3 +201,55 @@ pub async fn query(conn: &mut Connection, stmt: &str, options: &Options)
     }
     Ok(())
 }
+
+fn truncated(s: &str, max_chars: usize) -> &str {
+    match s.char_indices().nth(max_chars) {
+        None => s,
+        Some((idx, _)) => &s[..idx],
+    }
+}
+
+async fn interpret_file(file: String, options: Options)
+    -> Result<(), anyhow::Error>
+{
+    let mut contents = async_read(file).await?;
+    contents.push(b';');
+
+    let mut input = vec![];
+
+    loop {
+        match full_statement(&contents, None) {
+            Ok(len) => {
+                let stmt = String::from_utf8((&contents[..len]).to_vec())
```

Comment:
I tried doing that before and this is what the compiler says:

```
error[E0506]: cannot assign to `contents` because it is borrowed
   --> src/non_interactive.rs:228:17
    |
223 |                 let stmt = str::from_utf8(&contents[..len])?.trim();
    |                                            -------- borrow of `contents` occurs here
224 |                 if stmt != "" && stmt != ";" {
225 |                     input.push(stmt);
    |                     ----- borrow later used here
...
228 |                 contents = (&contents[len..]).to_vec();
    |                 ^^^^^^^^ assignment to borrowed `contents` occurs here

error[E0505]: cannot move out of `contents` because it is borrowed
   --> src/non_interactive.rs:234:51
    |
223 |                 let stmt = str::from_utf8(&contents[..len])?.trim();
    |                                            -------- borrow of `contents` occurs here
...
234 |                 let remainder = String::from_utf8(contents)
    |                                                   ^^^^^^^^ move out of `contents` occurs here
...
249 |     for q in input {
    |              ----- borrow later used here
```

The current code is my attempt to evade this error.

Reply:
Ah, that `contents[len..].to_vec()` is another too slow allocation that will give you `O(n^2)` (where n == size of file) performance. What you want to do is have a slice and always operate on it:
```rust
    let mut contents_buf = async_read(file).await?;
    contents_buf.push(b';');
    let mut contents = &contents_buf[..];
    // Later you can freely reassign `contents` (but not buf):
    contents = &contents[start..];
```

Reply:
Got it. That code is gone though, but this is a good suggestion, next time I'll try to get it right.